wallet: don't duplicate change output if already exist

furszy opened the pull request:
If the transaction includes an output that goes to the custom change
address (`CoinControl::destChange`), the transaction creation process
should avoid creating duplicate outputs to the same address.

Instead, it should reuse the existing output to prevent revealing which
address is the change address, which could compromise privacy.

This will also be useful to make other wallet processes less confusing
and error-prone. For instance, within the bumpfee functionality, we
manually extract the change recipient from the transaction before calling
'CreateTransaction' to ensure that the change output is not duplicated.

Side note:
This is something that will be using for the #26732 rework.

furszy committed:
wallet: FundTransaction, don't add change as another recipient if exists

If the transaction includes an output that goes to the custom change
address (CoinControl::destChange), the transaction creation process
should avoid creating duplicate outputs to the same address.

Instead, it should reuse the existing output to prevent revealing
which address is the change address, which could compromise privacy.

This will also be useful to make other wallet processes less
confusing and error-prone. For instance, within the bumpfee
functionality, we are extracting the change recipient from the
transaction before calling 'CreateTransaction' to ensure that the
change output is not duplicated.

furszy committed:
test: fundrawtxn, add coverage for double change output creation

Coverage for fundrawtxn:

1) Should not create a second change output if there is
   one already going to that address.

2) The wallet changeless predilection should never discard
   a manually set output by the user.
   e.g. the user manually crafted a tx that already contain a
   change output with certain value. This output shouldn't get
   lost as it was predefined by the user. Only should be used
   as recipient of the "inputs - outputs" surplus during
   'fundrawtransaction'.

furszy committed:
wallet: refactor, move recipients sum loop after change script creation

this will be used in the follow-up commit to detect whether one of the
recipients matches the change script or not.

Xekyo commented on src/wallet/spend.cpp:
```diff
@@ -884,6 +874,19 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
         CHECK_NONFATAL(IsValidDestination(dest) != scriptChange.empty());
     }
+
+    // Gather total out and whether SFFO is enabled or not
```
```suggestion
    // Gather sum of recipient output amounts and count number of outputs to subtract fees from
```

furszy replied:
done

Xekyo commented on src/wallet/spend.cpp:
```diff
@@ -987,14 +991,21 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
-            return util::Error{_("Transaction change output index out of range")};
+        if (existent_change_out_index) {
+            // If there is an output with the change address already, increase that one instead of creating a new output
+            txNew.vout.at(*existent_change_out_index).nValue += change_amount;
```
I would expect `GetChange(…)` to deduct the fee for the change, but if you already have a change output, wouldn’t that mean that you account for the change fee twice? You might need to add the change_fee back in the case where you just increase an existing change output.

furszy replied:
It depends on how the output was crafted. If it was created manually, it will not account for the fee. And also, it will not account for it if it was created in a previous `createTransaction` call with SFFO enabled.

It is fine to "overpay" the fee at this line because later in the process (50 lines below) we check if the final fee is above the needed one, automatically increasing the change by the surplus if needed.

S3RK replied:
Isn't the fee for all outputs accounted for in `not_input_fees`? It seems to me this will always overpay and then will be corrected when we detect surplus fee.

Maybe a clearer way to handle this would be to pass `chnage_fee=0` to `GetChange` if `existent_change_out_index=true`. I'd prefer to minimise number of cases when we need to correct fees at the later stage of transaction building.

Xekyo commented on test/functional/wallet_fundrawtransaction.py:
```diff
@@ -315,6 +316,67 @@ def test_change_type(self):
+                # Non-change output must remain the same
+                assert_equal(out['value'], outputs[external_addr])
+                assert_equal(out['scriptPubKey']['address'], external_addr)
+
```
Perhaps you should add a test that checks how much fees you’ve paid for the change outputs. I suspect you might pay two change fees when increasing an existing change, unless that gets cleaned up in some fashion. A test that rules it out would be neat.

furszy replied:
Yeah, it gets cleaned at the end of the tx creation process (check https://github.com/bitcoin/bitcoin/pull/27601#discussion_r1247203254).

But new tests never hurt, so just added it inside e46c1276. Thanks for pushing me to do it.

murchandamus reviewed (commented):
Concept ACK

luke-jr commented:
Concept ACK

furszy reviewed (commented):
Updated per feedback, thanks @Xekyo.

Added test coverage verifying that the same fee is paid when the wallet creates a change output automatically vs when the user provides the change output and the wallet just use it as recipient for the fee surplus.

S3RK commented:
How would it work if coinselection produced result without change output?

furszy reviewed (commented):
> How would it work if coinselection produced result without change output?

There is a test exercising the behavior inside a8ac2ad8 and also an explanation there.

But essentially, the process must adhere to the outputs specified by the user. If the change destination option in 'fundrawtransaction' matches one of the output scripts, the wallet is instructed to 'increase the specified output only if there is any change' and must not discard outputs predefined by the user under any circumstances

furszy committed:
refactor: remove change_spend_size field from CoinSelectionParams

The field is an intermediate result used to calculate other values
and not read anywhere inside coin selection.

furszy committed:
coinselection: option to exclude change cost from waste score

Useful when the transaction already accounted for the change cost.
Which happens when the user, or wallet process, set the coin control
custom change address equal to one of the existing output addresses.

In other words, when the user selects one of the transaction existing
outputs to be increased by the difference between inputs - output.

This will be connected to the transaction creation process in the
following-up commit.

furszy committed:
refactor: wallet spend, decouple change params initialization

No behavior change.

The initialization of all the CoinSelectionParams fields related
to 'change' were moved to its own standalone function.

furszy committed:
wallet: compute change params only when we might produce change output

Instructing coin selection to not account for the 'change' costs.

For example:

* The minimum viable change will be 0, which means that the surplus
  between inputs and outputs will never be sent to fees.

* With change_fee=0:
  - The target value being sought will be the exact target (and not
     target + change_fee).
  - 'GetChange' will not deduce the change_fee from the value
     calculation.

* The waste score calculation will exclude the change cost.

furszy committed:
refactor: remove discard_feerate field from CoinSelectionParams

The field is an intermediate result used to calculate other values
and not read anywhere inside coin selection.

furszy reviewed (commented):
Updated.  Added required changes to make coin selection, change and fee calculations take into consideration the presence of an existing change output. Avoiding cases where we end up having to increase the change output at the end of the process by the fee surplus.

Changes:
* Cleaned two change related members from the `CoinSelectedParams` struct that are not used inside the coin selection process. They were only intermediate results used to calculate other values.
* Excluded the change cost from the waste score when the transaction already accounts for it.
* In transactions with an existing change output, the spending process will no longer compute all the change related coin selection params. By setting them to 0, we instruct coin selection to produce results closer to the target.

S3RK commented on src/wallet/spend.cpp:
```diff
@@ -919,21 +916,24 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
     // So cost of change = (change output size * effective feerate) + (size of spending change output * discard feerate)
     coin_selection_params.m_change_fee = coin_selection_params.m_effective_feerate.GetFee(coin_selection_params.change_output_size);
-    coin_selection_params.m_cost_of_change = coin_selection_params.m_discard_feerate.GetFee(change_spend_size) + coin_selection_params.m_change_fee;
+    coin_selection_params.m_cost_of_change = discard_feerate.GetFee(change_spend_size) + coin_selection_params.m_change_fee;
```
in 9dc5bbe042f4d1febde8753a8c56da35eedd9d3e

I know this preserves current behaviour, but don't we want to use `long_term_fee_rate` to determine cost of chance?

@murchandamus

S3RK commented on src/wallet/spend.cpp:
```diff
@@ -819,6 +819,41 @@ static void DiscourageFeeSniping(CMutableTransaction& tx, FastRandomContext& rng
     }
 }
 
+static void ComputeChangeParams(CoinSelectionParams& coin_selection_params, CWallet& wallet, const CScript& scriptChange, CAmount recipients_sum, size_t recipients_size)
```
in ea57562b4ab4071fd1a716f09e13d19ff68e99f1

I see how it's useful for the next commit, but the contract of the method is not really clear without reading the code.

furszy replied:
Guess that you are referring to the last two args?

I think that it isn't much of a problem because it is an internal static helper function.

They are needed for Knapsack min change target calculation. What maybe could do is to pass them as one: `mean_amount = std::floor(recipients_sum / recipients_size)` and document the function properly.

And also, in a follow-up, we could go further and move the else branch of the `scriptChange` block of code inside `ComputeChangeParams` too. That part is only needed when there isn't an existent change output.

S3RK replied:
Let's document which part of `CoinSelectionParams` is modified and which is not.

S3RK replied:
I mean `ComputeChangeParams` computes some of `CoinSelectionParams`, but it's not immediately obvious which ones without reading the code of the function

furszy replied:
> I mean `ComputeChangeParams` computes some of `CoinSelectionParams`, but it's not immediately obvious which ones without reading the code of the function

Ok. Added docs.
The function calculates all the `CoinSelectionParams` fields related to 'change'. No exceptions.

S3RK commented:
Approach ACK

I think setting change related params to 0 in case where we reuse existing output is the correct way to go.

furszy reviewed (commented):
Updated per feedback. Thanks.

Added docs for the `ComputeChangeParams` new function.

achow101 commented:
While I agree that doing something smart like this in the context of fee bumping is fine, `CreateTransaction` is a generic function for the wallet's transaction building so changing the behavior here will have an effect on other uses as well, so I don't think this is the right approach. In general, we shouldn't assume what the change is when it is ambiguous - conceivably a user could be reusing addresses or otherwise actually wants a particular change address that is also one of the outputs. Doing this is allowed now, and I don't think we should change the meaning of the change address argument.

> Instead, it should reuse the existing output to prevent revealing which
address is the change address, which could compromise privacy.

No, reusing the output reveals the change since the amount changes.

murchandamus commented:
Perhaps this is a good way forward: only merge the new change output and the existing output on the original transaction if either a) Bitcoin Core organically detects that the original output was the change and it is also provided as the custom change address again, or b) if the user explicitly instructs the wallet to treat the original transaction’s output as the change per https://github.com/bitcoin/bitcoin/pull/26467 and provides the same address as custom change again. Otherwise, if the output script associated with the custom change address  matches an existing output on the original transaction, it should fail and request more precise instructions from the user.

Other than that, I agree that a surveillant seeing the same output script in the original and replacement transaction but with different amounts would have a strong hint that it’s the change. Two outputs to the same output script would also be suspicious, especially if the tx fingerprint matches Bitcoin Core, since it does not permit duplicate outputs to the same output script usually.

furszy reviewed (commented):
> In general, we shouldn't assume what the change is when it is ambiguous - conceivably a user could be reusing addresses or otherwise actually wants a particular change address that is also one of the outputs. Doing this is allowed now, and I don't think we should change the meaning of the change address argument.

That is not really how the wallet behaves for regular outputs. Not sure why it should
behave differently for change outputs only.

Right now, the user cannot duplicate outputs in any of the wallet related RPC commands.
They are all blocking dup output scripts (Check `ConstructTransaction`, inside `AddOutputs`).
Only the change output is auto-magically duplicated by the wallet without the user consent.

The only way for users to duplicate outputs destinations is to use the `createrawtransaction`
command. Which I think that is good, the wallet RPC related commands should be "safe" to
use (they shouldn't duplicate outputs automatically nor do stuff that can leak information),
and then the raw creation commands should be rules free for people to do whatever they want.


> > Instead, it should reuse the existing output to prevent revealing which
> > address is the change address, which could compromise privacy.
> 
> No, reusing the output reveals the change since the amount changes.

Depends on the context. If this is used to bump certain transaction, the outcome is
practically the same for both. Either increasing the single change output or creating another
output with the same script compromises privacy (more if we are the only software doing it).

But, for transactions being funded (not published), it is worst to create a tx with
two outputs going to the same script rather than increasing the amount of one of them.

---------------------------------------

I think that if your concern is about allowing the user to have a second output going to
the same change address, they could also do it with this as well. They just need to
craft a raw transaction with two outputs going to the change address. The wallet
will take one of them as change, and increase it.

Still, will think more about this. I'm feeling that you surely have a point (other than
the ones written above) but I'm not seeing it yet. It feels that it is related to a
"use existing change output if exist" option instead of automatically using one of
the outputs when it matches to the user provided change destination.

But.. this last thing also sound weird. The user provided the outputs, and also set
a custom change destination that matches one of them. So.. it should be more than
fine to interpret it as "use this output for change".

furszy reviewed (commented):
Rebased due a silent merge conflict.

achow101 commented:
> That is not really how the wallet behaves for regular outputs. Not sure why it should
> behave differently for change outputs only.

Then it should just fail instead of merging the outputs. It should tell the user that they've done something wrong, as we do in the RPCs. I agree with @murchandamus's suggestion.

> The only way for users to duplicate outputs destinations is to use the `createrawtransaction`
> command.

No, that forbids duplicates too.

> But.. this last thing also sound weird. The user provided the outputs, and also set
> a custom change destination that matches one of them. So.. it should be more than
> fine to interpret it as "use this output for change".

I think the issue is mainly that this is not expected typical user behavior. We don't expect users to have an address as both a normal output and as the change address, in the same way that we don't expect users to have the same address twice. Because this is unexpected of our users, we shouldn't make assumptions about why they may be doing this and should instead tell them that they've done something unexpected.

furszy reviewed (commented):
Thanks achow for the feedback :).


> > That is not really how the wallet behaves for regular outputs. Not sure why it should
> > behave differently for change outputs only.
> 
> Then it should just fail instead of merging the outputs. It should tell the user that they've done something wrong, as we do in the RPCs. I agree with @murchandamus's suggestion.

Ok, great. We are almost on the same page. There is one point about Murch's suggestion that makes some noise in my head.

#### Describing the general process first:
When the user manually specifies a custom change address/script, the transaction creation and transaction funding processes are instructed to follow one of the following option:
1) If there is an existing output matching the user provided change destination: the process will send the change amount to the existing output.
2) If there is no output matching the user provided change destination: the process will create a new output under the provided destination and send the change amount to it.

#### Murch's suggestion:
The only difference between the general process description and what murch suggested is the following one; he said to only perform (1) when the wallet organically detect the custom change address provided by the user as change (which makes the existing output matching this address also change). In other words, he said to only perform (1) when the wallet's `IsChange()` returns true for the provided destination.

#### Now, the discussion point.
What is the point behind forbidding this action? Does it have any advantage over the current status?

If the concerns are related to sending the change amount to an address not owned by the wallet; that is already possible today. Simply by specifying a custom change address, the user can send the change amount to any destination. This is part of the contract behind the utilization of the custom change address option.

My perspective is that what this PR introduces isn't radically different from our current process. The only difference arises when one of the user's manually specified outputs matches the user's manually specified change address. In such cases, we will increase the existing output instead of creating a second output going to the same address.

I mean, I feel that what we are introducing here is a natural behavior rather than something unexpected. As a user, if I provide a custom change address that relates to one of the outputs that I provided, it suggests my intention to send the change to the existing output. I don't think that anyone would expect another behavior.

But, having that said, I agree with you that we should explicitly state this behavior on the RPC custom change address option description. So there are no doubts about it.

Would you agree with this rationale, or do you still have second thoughts about it?

S3RK commented:
> We don't expect users to have an address as both a normal output and as the change address, in the same way that we don't expect users to have the same address twice. 

I think there is a difference in this two cases.
a) When user provided the same address twice and we show an error, then the user can easily achieve the intended goal by just summing the amounts. Having two UTXOs don't give an advantage over a one with combined amount as spending conditions are the same anyway, so there is no point.
b) If user provided the same address for output and change and we show an error. What the user should do? Pick a different change address? But new change address will have different spending conditions, so this is not an equivalent substitute. And even then, it's more expensive as it forces an extra output. I could be wrong, but there is no way for the user to specify they want to send change to an existing output

Sjors commented:
For the `send` RPC at least (have to think about the others) I think that if a user sets the same address in `outputs` and for `change_address` that should be considered an error. They should either remove it from outputs, or use the `change_position` argument.

achow101 commented:
> I mean, I feel that what we are introducing here is a natural behavior rather than something unexpected. As a user, if I provide a custom change address that relates to one of the outputs that I provided, it suggests my intention to send the change to the existing output. I don't think that anyone would expect another behavior.

I disagree. I think it's a mistake if that happens. If the user wants to use something for change, then they shouldn't be giving it some value first, because it's change, not an recipient.

On second thought, I think we should actually just always error, regardless of whether the wallet thinks the address is change.

> If user provided the same address for output and change and we show an error. What the user should do? Pick a different change address?

Either pick a different change address, or remove the output.

> I could be wrong, but there is no way for the user to specify they want to send change to an existing output

There isn't, but it also doesn't make sense to me for someone to want to do that. Why specify an existing amount instead of just dropping it and letting the software add one if it has to? Maybe coin selection finds a solution that doesn't have change and the output doesn't actually need to exist.

S3RK commented:
> There isn't, but it also doesn't make sense to me for someone to want to do that. Why specify an existing amount instead of just dropping it and letting the software add one if it has to? Maybe coin selection finds a solution that doesn't have change and the output doesn't actually need to exist.

I think this is the crux of the discussion here, and we should agree on the use-case before talking about implementation.

Specifying one of the existing outputs as change would be useful when a user wants to pay someone and send to themselves (at least X sats) in the same tx. I can think of following examples:
* Consolidating UTXOs and adding a low priority payment to it.
* Moving funds between wallets and paying someone.

What do you think? From your experience, does it sound like realistic use-case?

murchandamus commented:
> Specifying one of the existing outputs as change would be useful when a user wants to pay someone and send to themselves (at least X sats) in the same tx. I can think of following examples:
> 
>     * Consolidating UTXOs and adding a low priority payment to it.
> 
>     * Moving funds between wallets and paying someone.
> 
> What do you think? From your experience, does it sound like realistic use-case?

Yeah those sound like things that I would consider doing if I were operating a high-volume wallet, e.g. in a multi-wallet setup the receive-only wallet could send change to the cold-wallet when topping up the send-only withdrawal wallet.

murchandamus commented on src/wallet/spend.cpp:
```diff
@@ -946,8 +946,10 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
-    // Set the 'change' related coin selection params
-    ComputeChangeParams(coin_selection_params, wallet, scriptChange, recipients_sum, vecSend.size());
+    // Only compute 'change' params when we might produce a change output
+    if (!existent_change_out_index) {
+        ComputeChangeParams(coin_selection_params, wallet, scriptChange, recipients_sum, vecSend.size());
+    }
```
No, please don’t do this. This is hacky and muddles meaning of variables.

The parameters of creating change are independent of whether change is created or not. In fact, we use some of these value in some of the coinselection algorithms to determine whether we have a changeless solution as well as in transaction building to determine the size of the change output. By fiddling with these values, it feels like we are coding against the implementation of the `GetChange(…)` function rather than against the function definition. Please use a new designated variable to explicitly track that transaction building should not create a change output if we absolutely must have that.

I would much prefer an approach that sidesteps this entire issue by starting the creation of an RBF transaction higher in the stack where we can remove a pre-existing identifiable change output from the draft transaction before just kicking off an entire transaction building process with some preset inputs.

murchandamus replied:
This is also related to https://github.com/bitcoin/bitcoin/pull/28366 (last two commits only, rest is #26152).

murchandamus reviewed (commented):
Concept ACK, but I dislike the approach:

furszy commented:
> Are you still working on this?

Probably next month. But no problem on closing it for now. Will re-open it or create a new one whenever I back to it.
